runhouse/resources/hardware/cluster.py

```diff
@@ -787,6 +815,15 @@ def restart_server(
             self.client.use_https = https_flag
             self.client.cert_path = self.cert_config.cert_path
 
+        if restart_ray:
+            # Restart Ray on the head node and each of the workers
+            # TODO: kill ray on all nodes first. Need to think more of the
```

Comment:
We should find out if it dies automatically on the workers when we kill it on the head or if we're really borking them when we pkill the head node.

Reply:
Seems like 'RAY_HEAD_IP=127.0.0.1 RAY_HEAD_PORT=6379 ray stop' should stop the entire cluster with workers when run on the head node.

Reply:
Nope, it doesn't seem to work..

Back to pkill for now. 

The workers do not die automatically, they become orphaned. There seems to be a way to carefully send shutdown commands to them, but need to experiment further. We can probably decouple it from the current PR?

Reply:
Next to try: python3 -m "import ray; ray.init('ray://localhost:6379'); ray.shutdown()"
https://docs.ray.io/en/latest/ray-core/api/doc/ray.shutdown.html

If that doesn't work, maybe this is a more involved and explicit way:
https://stackoverflow.com/questions/69613739/how-to-kill-ray-tasks-when-the-driver-is-dead